### ISSUE TITLE
Support sub-stream processing

### ISSUE THREAD

**Issue by hyperbolic2346** [Motion-Project/motion#123]
IP cameras keep getting larger sensor and in order to keep up we will need to be able to do motion detection on the sub-stream since it is a lower resolution and then do the recording on the main stream.

**Comment by osmaa** [comment]
Has anyone considered doing motion detection from the motion vector data of an already-compressed video stream? I've been experimenting with using H264 motion vector data and it works really nicely. Such a recorder wouldn't necessarily have to process the full-resolution video at all, unless annotations like motion boxes are desired.

Related research: https://www.researchgate.net/publication/221315082

**Comment by Mr-Dave** [comment]
I currently do the substream processing like this issue is requesting manually via calls to scripts from the Motion event triggers.  I've also seen it within other forks so I am thinking that this will be could be completed as an enhancement in the future.  With respect to the comment by osmaa, the research article points out an ideal method but I think it would be an extensive effort that would need to go into a different topic.  At this time, I think we'd need to find a sample implementation of the research methods before being considered.

**Comment by hyperbolic2346** [comment]
@Mr-Dave I never thought to do it via a trigger script. I assume this makes it near impossible to do pre-record and things like that, but is there a noticeable delay to start recording the primary stream?

@osmaa That is very interesting and I'd love to see a test of that implemented.

**Comment by Mr-Dave** [comment]
Yes, pre-record or adding the extra timestamps, etc isn't possible.  I miss the first second or two but that isn't what I'm worried about with my capturing.  The videos are much better with higher res/framerates as well as using the native compression.  I also set it to just record for a fixed amount of time and have the videos overlap so I don't miss anything during an extended event.  e.g.  Set the record time for the video to 30sec and the Motion gap to 25.    This is the line from the script ($1 is the time to record.  Others should be straight forward). <code>ffmpeg -rtsp_transport tcp -loglevel warning -i $NETCAMURL -c copy -t $1 $OUTPUTFILE1</code>

**Comment by osmaa** [comment]
It's a bit trickier to parse the h264 stream off a webcam to re-extract the motion vectors, but I've been playing with the Raspberry Pi camera, the MMAL interface provides direct access to the MV data. There are a few examples of that, while my own is still in bit of development, eg this one: https://gist.github.com/spikedrba/aecb82d8b51e991dbd01

The nice aspect of this approach is that even on the relatively underpowered original RPi Model B (single-core ARMv6), I am able to capture even pre-movement frames by keeping a circular memory buffer of the H264 frames which get dumped to file only if the motion vector data (which is lightweight enough to analyse in a python script!) reveals movement in the stream -- thanks to the heavy lifting being done by the SoC's own H264 encoder.

**Comment by ektus** [comment]
The vector detection from the MMAL interface works great, as can be experienced with pikrellcam (https://billw2.github.io/pikrellcam/pikrellcam.html). I've got it running in parallel with Motioneye on a RPi2, and use Motioneye on the preview stream of a megapixel-class RTSP camera. The detection quality of Pikrellcam is way better, even without much tweaking. The main stream of the RTSP camera is recorded 24/7 on a NAS to retain the high-res pictures for further analysis, should the need arise.

I'd vote for both improvements, motion detection on the preview stream greatly reduces CPU load, but so does vector detection by the GPU or camera and even combines it with better detection results.

**Comment by tosiara** [comment]
I also would like to note, that dumping raw stream without decoding and re-encoding will create a video file without time stamp and location box. And as devices get more power and dedicated h264 encoding chips - the issue with secondary stream may go away itself

FWIW. Just for the reference, with `extpipe` you can offload encoding to other cpu core(s). I was able to set up 1920x1080@30fps MJPEG UVC camera with ODROID-C1 and successfully had been encoding video files on remaining 3 cores. Entire system load was about 80% and all worked perfectly. If ARM computing power is going to grow twice in the near few years - you will be able to process two 1080 cameras or maybe one 2k

**Comment by ektus** [comment]
But in two years I'd probably want to replace all of my cameras with 2K and beyond, and would have to still use one PI for every camera 'cause the data rate grows as fast as the processing capability. It's intelligent software design that has to come to play, not raw computing power.

With RTSP cameras simultaneously delivering a high-res and a low-res stream, the scaling is offloaded to the camera and lessens load on the picture computing device, as is the topic with this enhancement issue.

**Comment by lolouk44** [comment]
> It's intelligent software design that has to come to play, not raw computing power.

I would second ektus' view / approach,

**Comment by tosiara** [comment]
If you have enough compute power on your motion box - then you don't care if your camera is able to give you the second lowres stream. IMHO, in long term investment into code change to support second stream (to offload stream resize to camera's hardware chip) would be higher than just increasing resources.

Has anyone tried utilizing RPi hardware decoding chip with motion? Licensing issue?

**Comment by hyperbolic2346** [comment]
I would say that the code changes required for supporting the sub-stream wouldn't be terrible. I do wonder when we hit the point of pain with the codebase though. It really is designed with a loop that wants an image to process and then write that image to the disk if there is motion. We might need to rethink some of this architecture in a new world where the motion detection might be run from a different stream or even a completely different method like the h264 motion data. I can see applications other than the substream as well. Like my security cameras record these x cameras when any of them detect motion or something like that.

I put this up as a nice to have, but honestly thought that I would eventually get to this. I'm also most interested in pass-through recording of the main stream, though it could mean giving up the box around motion and the timestamp. I was going to look at somehow muxing that into the video though internally, but we'll see I guess.

**Comment by lolouk44** [comment]
Hi Mr-Dave,

I have a short question about your script
`ffmpeg -rtsp_transport tcp -loglevel warning -i $NETCAMURL -c copy -t $1 $OUTPUTFILE`

I get how you'll record 30 sec at `on_event_start`.
If the event length is shorter than 30 sec, you still get 30 sec of video, no issues here
What happens if the event is longer than 30 sec? I understood the `event_gap` to be a number of sec of no motion before ending the event.
I can see how to set a `max_movie_time` but I understand this is not related to the event.
Where do I set an event max time to be 25 sec?
or am I missing something?

**Comment by Mr-Dave** [comment]
max_movie_time is effectively the same thing as a maximum event time.  Use the max_movie_time = 25.

**Comment by lolouk44** [comment]
Thanks :)

**Comment by iskunk** [comment]
Isn't the short delay in opening the primary stream a showstopper? If you detect motion on the sub-stream, and your recording of the primary stream starts two seconds later, you could very well miss whatever caused the motion in the first place. (Ultimate CCTV hack: run past the cameras)

As a workaround, you could have both streams going in parallel at all times. That would allow pre-record, but is rather harder on the camera/network (assuming the camera can handle it) and arguably worse than just pulling the primary stream.

Wouldn't it make more sense to just pull the primary stream, scale it down, and run motion detection on the smaller image? There should be good options for fast 2D scaling, and you can continue to do pre-record as always.

Re motion-vector detection (@osmaa): IMO, this is the way of the future. I actually obtained a PDF copy of that paper through my institution, and the idea seems no worse code-complexity-wise than what Motion is already doing. But there are multiple architectural issues that will need to be addressed before this sort of thing can be integrated.

As a starting point, FFmpeg provides an example program [extract_mvs.c](https://github.com/FFmpeg/FFmpeg/blob/master/doc/examples/extract_mvs.c) that shows how to get motion-vector data. It's fairly raw, and there are [many questions online](https://www.google.com/search?q=ffmpeg+extract_mvs.c) about this code (I'm wondering if this is H.264-only, of if e.g. VP9 would also work), but at least there's some concrete info there.

(Incidentally, there are a whole slew of other papers that address this topic area. I downloaded a second one that proposes a codec-independent approach involving statistical analysis of the opaque stream content. You could spend all day browsing these.)

Re adding timestamps, location boxes, etc.: Couldn't this be done by adding a parallel stream, as subtitles are sometimes done? From [this page](https://en.wikibooks.org/wiki/FFMPEG_An_Intermediate_Guide/subtitle_options):

> Subtitles may come either from a separate text file (typically with a .srt extension) or from a dedicated stream within the video file. In the latter case, the subtitles are typically stored not as text but as bitmaps that are overlayed on the video and may be referred to as "VOBSUBs" for DVDs or "PGS" for Bluray.

As the stream is just an overlayed bitmap, it could include text, boxes, whatever.

**Comment by jogu** [comment]
Yes, I think to achieve as good as we have as the moment, we'd need to be continually streaming both the low and "normal" streams. It would take at least a second or two to start recording a new stream.

Cameras are generally quite happy streaming both streams at the same time in my experience.

**Comment by iskunk** [comment]
That reduces CPU usage at the cost of increased network traffic (and the complexity of dealing with two streams from a camera instead of one). But consider that this particular trade-off usually goes the other way, where burning CPU is preferable to sending more packets (for example, HTTP compression, or even `*.tar.xz` files).

That's part of the beauty of the motion-vector approach, IMO: the cameras can go to 5MP and beyond, but the CPU usage doesn't have to increase to the same degree. Analysis can continue to be done on the primary stream without needing monster CPUs, and that should remain the case for years to come.

**Comment by hellyberry** [comment]
This enables motion for even a smaller device like the Raspberry Pi to process maximal number of web- or ip-cameras connected. This would allow optimizing resource usage to what one has, e.g. if bandwidth is available.

**Comment by Mr-Dave** [comment]
For those of you interested, the following is the latest branch that implements the dual stream processing.  In my testing, I have not seen a material decrease in CPU usage as it appears that the majority of the CPU usage is with the capture and decode rather than the motion detection.

https://github.com/Mr-DaveDev/motion/tree/high-resolution

**Comment by iskunk** [comment]
@Mr-Dave, could you investigate further that split of CPU usage? That goes completely against what I understood to be the relative complexity of motion detection, and would have implications for the desirability of e.g. motion-vector-based detection.

(Were you testing with a fairly high-res primary stream?)

**Comment by hyperbolic2346** [comment]
I feel the same way. I'm pretty certain all my cpu time is in the motion
detection, but I don't have anything to back that up yet.

**Comment by Mr-Dave** [comment]
The following is my graph of CPU usage.  I was surprised as well and why I posted the update to see if others want to test and validate the result. 

Single stream@320, dual stream@320&1920, single stream@1920, dual stream with my pass-though that copies packets but bypasses decoding the high resolution stream.
![image](https://user-images.githubusercontent.com/5959350/31644778-4e202a52-b2b5-11e7-850f-1978a1998a3d.png)

**Comment by iskunk** [comment]
I can't help but wonder if maybe you've got detection configured to do less, or are linking against an unoptimized FFmpeg, etc. Every time I've done a profile run with detection running, `alg_update_reference_frame()` is the heaviest function by far. Something's not right here...

**Comment by tosiara** [comment]
FWIW, there was NEON/SSE optimization proposed for the "heaviest" function `alg_update_reference_frame`
https://github.com/Motion-Project/motion/pull/498

**Comment by Mr-Dave** [comment]
Closed via #522
